ush/python/pygfs/task/archive.py

```diff
@@ -417,3 +438,151 @@ def replace_string_from_to_file(filename_in, filename_out, search_str, replace_s
+        freq = arch_dict.ARCH_EXPDIR_FREQ * 3600
+
+        # Skip gfs and enkf cycled RUNs (only archive during gdas RUNs)
+        # (do not skip forecast-only, regardless of RUN)
+        if arch_dict.NET == "gfs" and arch_dict.MODE == "cycled" and arch_dict.RUN != "gdas":
+            return False
+
+        # Determine if we should skip this cycle
+        # If the frequency is set to 0, only run on sdate (+assim_freq for cycled) and edate
+        if freq == 0:
+            if mode == "forecast-only" and (current_cycle != sdate or current_cycle != edate):
+                return False
+            elif mode == "cycled" and (current_cycle != sdate + assim_freq or current_cycle != edate):
+                return False
+        # Otherwise, the frequency is in hours
+        elif mode == "forecast-only" and (sdate - current_cycle).total_seconds() % freq != 0:
+            return False
+        elif mode == "cycled" and (sdate + assim_freq - current_cycle).total_seconds() % freq != 0:
+            return False
+
+        # Looks like we are archiving the EXPDIR
+        return True
+
+    @staticmethod
+    @logit(logger)
+    def _pop_git_info(arch_dict: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        This function checks the configuration options ARCH_HASHES and ARCH_DIFFS
+        and ARCH_EXPDIR_FREQ to determine if the git hashes and/or diffs should be
+        added to the EXPDIR for archiving and execute the commands.  The hashes and
+        diffs will be stored in EXPDIR/git_info.log.
+
+        Parameters
+        ----------
+        arch_dict: Dict
+            Dictionary with required parameters, including the following:
+
+            EXPDIR: str
+                Location of the EXPDIR
+            HOMEgfs: str
+                Location of the HOMEgfs (the global workflow)
+            ARCH_HASHES: bool
+                Whether to archive git hashes of the workflow and submodules
+            ARCH_DIFFS: bool
+                Whether to archive git diffs of the workflow and submodules
+        """
+
+        # Get commonly used variables
+        arch_hashes = arch_dict.ARCH_HASHES
+        arch_diffs = arch_dict.ARCH_DIFFS
+        homegfs = arch_dict.HOMEgfs
+        expdir = arch_dict.EXPDIR
+
+        # Find the git command
+        git = which('git')
+        if git is None:
+            raise FileNotFoundError("FATAL ERROR: the git command could not be found!")
+
+        output = ""
+        # Navigate to HOMEgfs to run the git commands
+        with chdir(homegfs):
+
+            # Are we running git to get hashes?
+            if arch_hashes:
+                output += "Global workflow hash:\n"
+
+                try:
+                    output += git("rev-parse", "HEAD", output=str)
+                    output += "\nSubmodule hashes:\n"
+                    output += git("submodule", "status", output=str)
+                except ProcessError:
+                    raise OSError("FATAL ERROR Failed to run git")
```

Comment:
If we are going to catch and then raise new exceptions, we should use `from` to keep the info from the original (repeat for subsequent instances):
```suggestion
                except ProcessError as pe:
                    raise OSError("FATAL ERROR Failed to run git") from pe
```